RestCommClient/Classes/Push/PushHandler.m

```diff
@@ -0,0 +1,269 @@
+ * by the @authors tag.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation; either version 3 of
+ * the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ * For questions related to commercial use licensing, please contact sales@telestax.com.
+ */
+
+
+#import "PushHandler.h"
+#import "common.h"
+#import "PushCommunicationManager.h"
+
+//keys
+NSString *const kAccountSidKey = @"accountSidKey";
+NSString *const kClientSidKey = @"clientSidKey";
+NSString *const kApplicationSidKey = @"applicationSidKey";
+NSString *const kBindingSidKey = @"bindingSidKey";
+NSString *const kCredentialSidKey = @"credentialSidKey";
+
+@implementation PushHandler{
+    PushCommunicationManager *pushCommManager;
+    NSString *signalingUsername;
+    NSString *username;
+    NSString *password;
+    NSString *friendlyName;
+    NSString *rescommAccountEmail;
+    NSString *token;
+    BOOL sandbox;
+    
+    //certificate data directory
+    NSString *certificatePublicPath;
+    NSString *certificatePrivatePath;
+}
+
+- (id)initWithParameters:(NSDictionary *)parameters{
+    self = [super init];
+    if (self){
+        username = [parameters objectForKey:@"username"];
+        password = [parameters objectForKey:@"password"];
+        signalingUsername = [parameters objectForKey:@"signaling-username"];
+        friendlyName = [parameters objectForKey:@"friendly-name"];
+        certificatePublicPath = [parameters objectForKey:@"push-certificate-public-path"];
+        certificatePrivatePath = [parameters objectForKey:@"push-certificate-private-path"];
+        rescommAccountEmail = [parameters objectForKey:@"rescomm-account-email"];
+        token = [parameters objectForKey:@"token"];
+        sandbox = [parameters objectForKey:@"is-sandbox"];
+        
+        pushCommManager = [[PushCommunicationManager alloc] initWithUsername:username andPassword:password];
+    }
+    return self;
+}
+
+- (void)registerDevice{
+    if (!token || token.length == 0){
+        RCLogError("Push notification token is nil or empty.");
+        return;
+    }
+    
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        //before calling the pinding api, we need to get:
+        //account sid
+        //client sid
+        //application sid
+        //credentials sid
+        //binding sid
+        
+        //account id
+        [self getAccountSidWithCompletionHandler:^(NSString *accountSid){
```

Comment:
I'm worried that all the complexity introduced by all those block + completion handlers will come back to bite us in the future in terms of bugs and difficulty to troubleshoot issues. Could we maybe use synchronous http facilities and run on after the other instead, while all in a background thread of course so that we don't mess with UI thread? That could potentially make error reporting easier as well. Wdyt?

Reply:
I was start thinking in that way, but the only thing I have found, using the NSURLSession, is to use semaphore example: https://stackoverflow.com/questions/26784315/can-i-somehow-do-a-synchronous-http-request-via-nsurlsession-in-swift
And I saw "wait forever"

Reply:
You don't have to wait forever, you can use a timeout instead. The idea here is that we use tools that are fit for the job. Async is great in most cases, but right now I'm getting the impression that we are doing serial processing and with so many asynchronous operations + blocks we might be asking for trouble. In any case I don't feel that strongly about it; just want to make sure we know where we 're getting, so I'll let you decide

Reply:
I think we are good. I would like to leave it as it is.

Reply:
Ok, let's hope it doesn't come back to bite us